test: add runner watch mode isolation tests

pmarchini opened the pull request:
This PR extends the existing tests for the Node.js test runner's `--watch` mode by adding support for the `--experimental-test-isolation` option.

@cjihrig I was unable to find a test that covers this combination.  
I also noticed that the `should run new tests when a new file is created in the watched directory` test behaves differently depending on the isolation modes.  
What do you think?

cjihrig commented:
> What do you think?

Without having investigated the discrepancy, I'd say they should behave the same.

cjihrig commented on test/parallel/test-runner-watch-mode.mjs:
```diff
@@ -166,31 +168,38 @@ async function testWatch({
+        await testWatch({ fileToUpdate: 'test.js', action: 'rename', isolation });
+      });
+
+      it('should not throw when delete a watched test file', { skip: common.isAIX }, async () => {
```
We already skip the entire file at the very top when running on AIX.

pmarchini replied:
done

cjihrig commented on test/parallel/test-runner-watch-mode.mjs:
```diff
@@ -166,31 +168,38 @@ async function testWatch({
+        await testWatch({ fileToUpdate: 'test.js', action: 'delete', isolation });
+      });
+
+      if (isolation !== 'none') {
```
Instead of an if statement, we could make it a TODO or SKIP test. TODO will run but a failure will not fail the test suite. SKIP won't run at all.

pmarchini replied:
I left the comment because I was planning to fix the issue in this PR, but I haven't had the time in the last two days. I'm going to address your comments, and then I'll investigate the issue ASAP.
(Thanks, as always, for the feedback 😁)

pmarchini replied:
updated! 🚀

cjihrig reviewed (commented):
Mostly LGTM. Two small comments. Have you been able to identify why the one test fails without isolation?

RedYetiDev commented on test/parallel/test-runner-watch-mode.mjs:
```diff
@@ -166,31 +168,39 @@ async function testWatch({
+      });
+
+      it('should run new tests when a new file is created in the watched directory', {
+        todo: isolation === 'none' ?
+          'This test is failing when isolation is set to none and must be fixed' :
+          undefined,
```
Is this an issue with the test or with isolation?

If it's with isolation, can you link / open a specific issue for it?

pmarchini commented:
> Mostly LGTM. Two small comments. Have you been able to identify why the one test fails without isolation?

Hey @cjihrig, I just had the chance to take a look, and every time that a file changes (rename, delete, update), if we are in isolation `none`, then `restartTestFile` restarts all the tests without filtering based on the "added, updated, removed" logic.

I'll try to address this issue in a separate PR if you agree.

cjihrig commented:
That makes sense. Thanks for digging in. I'd say if you want to try working on it and think you can improve the experience then go for it! If not, it's still good to have these tests 😄

pmarchini commented:
> That makes sense. Thanks for digging in. I'd say if you want to try working on it and think you can improve the experience then go for it! If not, it's still good to have these tests 😄

I'll definitely give it a try 😁

mcollina reviewed (approved):
lgtm

RedYetiDev commented:
@pmarchini can you rebase from `main`?

RedYetiDev commented:
(Ignore my last [deleted] comment) I've added `commit-queue-squash` to this PR, it should be good to commit now.
